src/optimagic/visualization/backends.py

```diff
@@ -0,0 +1,131 @@
+import abc
+from typing import Any
+
+import plotly.express as px
+import plotly.graph_objects as go
+
+from optimagic.config import IS_MATPLOTLIB_INSTALLED
+from optimagic.exceptions import NotInstalledError
+from optimagic.visualization.plotting_utilities import LineData
+
+if IS_MATPLOTLIB_INSTALLED:
+    import matplotlib as mpl
+    import matplotlib.pyplot as plt
+
+
+class PlotBackend(abc.ABC):
+    default_template: str
+    default_palette: list
+
+    @abc.abstractmethod
+    def __init__(self, template: str | None):
+        if template is None:
+            template = self.default_template
+
+        self.template = template
+        self.figure: Any = None
+
+    @abc.abstractmethod
+    def add_lines(self, lines: list[LineData]) -> None:
+        pass
+
+    @abc.abstractmethod
+    def set_labels(self, xlabel: str | None = None, ylabel: str | None = None) -> None:
+        pass
+
+    @abc.abstractmethod
+    def set_legend_properties(self, legend_properties: dict[str, Any]) -> None:
+        pass
+
+
+class PlotlyBackend(PlotBackend):
+    default_template: str = "simple_white"
+    default_palette: list = px.colors.qualitative.Set2
+
+    def __init__(self, template: str | None):
+        super().__init__(template)
+        self._fig = go.Figure()
+        self._fig.update_layout(template=self.template)
+        self.figure = self._fig
+
+    def add_lines(self, lines: list[LineData]) -> None:
+        for line in lines:
+            trace = go.Scatter(
+                x=line.x,
+                y=line.y,
+                name=line.name,
+                mode="lines",
+                line_color=line.color,
+                showlegend=line.show_in_legend,
+                connectgaps=True,
+            )
+            self._fig.add_trace(trace)
+
+    def set_labels(self, xlabel: str | None = None, ylabel: str | None = None) -> None:
+        self._fig.update_layout(xaxis_title_text=xlabel, yaxis_title_text=ylabel)
+
+    def set_legend_properties(self, legend_properties: dict[str, Any]) -> None:
+        self._fig.update_layout(legend=legend_properties)
+
+
+if IS_MATPLOTLIB_INSTALLED:
+
+    class MatplotlibBackend(PlotBackend):
```

Comment:
If I experiment in a notebook, the matplotlib backend is shown twice, while the plotly figure is only shown once. Also if I assign the figure to a variable, plotly does not show anything (wanted behavior) while with the matplotlib backend the figure is still shown once.

This behavior should be aligned across backends, and we want the plotly bahavior.